code/modules/mob/living/basic/space_fauna/revenant/_revenant.dm

```diff
@@ -0,0 +1,499 @@
+	. += "Total Essence Stolen: [essence_accumulated] SE"
+	. += "Unused Stolen Essence: [essence_excess] SE"
+	. += "Perfect Souls Stolen: [perfectsouls]"
+
+/mob/living/basic/revenant/update_health_hud()
+	if(isnull(hud_used))
+		return
+
+	var/essencecolor = "#8F48C6"
+	if(essence > max_essence)
+		essencecolor = "#9A5ACB" //oh boy you've got a lot of essence
+	else if(essence <= 0)
+		essencecolor = "#1D2953" //oh jeez you're dying
+	hud_used.healths.maptext = MAPTEXT("<div align='center' valign='middle' style='position:relative; top:0px; left:6px'><font color='[essencecolor]'>[essence]E</font></div>")
+
+/mob/living/basic/revenant/say(message, bubble_type, list/spans = list(), sanitize = TRUE, datum/language/language = null, ignore_spam = FALSE, forced = null, filterproof = null, message_range = 7, datum/saymode/saymode = null)
+	if (!message)
+		return
+
+	if (src.client)
+		if(client.prefs.muted & MUTE_IC)
+			to_chat(src, span_boldwarning("You cannot send IC messages (muted)."))
+			return
+		if (!(ignore_spam || forced) && src.client.handle_spam_prevention(message, MUTE_IC))
+			return
+
+	if(sanitize)
+		message = trim(copytext_char(sanitize(message), 1, MAX_MESSAGE_LEN))
+
+	src.log_talk(message, LOG_SAY)
+	var/rendered = span_deadsay("<b>UNDEAD: [src]</b> says, \"[message]\"")
+	relay_to_list_and_observers(rendered, GLOB.revenant_relay_mobs, src)
+
+/mob/living/basic/revenant/ClickOn(atom/A, params) //revenants can't interact with the world directly, so we gotta do some wacky override stuff
+	var/list/modifiers = params2list(params)
+	if(LAZYACCESS(modifiers, SHIFT_CLICK))
+		ShiftClickOn(A)
+		return
+	if(LAZYACCESS(modifiers, ALT_CLICK))
+		AltClickNoInteract(src, A)
+		return
+	if(LAZYACCESS(modifiers, RIGHT_CLICK))
+		ranged_secondary_attack(A, modifiers)
+		return
+
+	if(ishuman(A) && in_range(src, A))
+		attempt_harvest(A)
+
+/mob/living/basic/revenant/ranged_secondary_attack(atom/target, modifiers)
+	if(revealed || inhibited || HAS_TRAIT(src, TRAIT_NO_TRANSFORM) || !Adjacent(target) || !incorporeal_move_check(target))
+		return
+
+	var/list/icon_dimensions = get_icon_dimensions(target.icon)
+	var/orbitsize = (icon_dimensions["width"] + icon_dimensions["height"]) * 0.5
+	orbitsize -= (orbitsize / world.icon_size) * (world.icon_size * 0.25)
+	orbit(target, orbitsize)
+
+/mob/living/basic/revenant/death()
+	if(!revealed || dormant) //Revenants cannot die if they aren't revealed //or are already dead
+		return
+	ADD_TRAIT(src, TRAIT_NO_TRANSFORM, REVENANT_STUNNED_TRAIT)
+	dormant = TRUE
+
+	visible_message(
+		span_warning("[src] lets out a waning screech as violet mist swirls around its dissolving body!"),
+		span_revendanger("NO! No... it's too late, you can feel your essence [pick("breaking apart", "drifting away")]..."),
+	)
+
+	revealed = TRUE
+	invisibility = 0
+	icon_state = "revenant_draining"
+	playsound(src, 'sound/effects/screech.ogg', 100, TRUE)
+
+	for(var/i = alpha, i > 0, i -= 10)
+		alpha = i
+		CHECK_TICK
```

Comment:
I don't think `check_tick` is the same thing as `stoplag()`

Reply:
```dm
/// runs stoplag if tick_usage is above the limit
#define CHECK_TICK ( TICK_CHECK ? stoplag() : 0 )
```

???

Reply:
But also you should just rework this to act less weird and buggy , it's just supposed to do a fade out to 0 alpha

Reply:
yeah i edited my previous comment upon thinking about it further, you're right